tendermint-testgen/src/helpers.rs

```diff
@@ -0,0 +1,83 @@
+//! Helper functions
+
+use serde::de::DeserializeOwned;
+use simple_error::*;
+use std::io::{self, Read};
+use tendermint::{vote, signature::Signature, amino_types};
+use signatory_dalek::Ed25519Verifier;
+use signatory::signature::Verifier;
+
+#[macro_export]
+macro_rules! set_option {
+    ($name:ident, $t:ty) => {
+        pub fn $name(mut self, $name: $t) -> Self {
+        self.$name = Some($name.clone());
+        self
+    }
+    };
+    ($name:ident, $t:ty, $val:expr) => {
+        pub fn $name(mut self, $name: $t) -> Self {
+        self.$name = $val;
+        self
+    }
+    };
+}
+
+/// Tries to parse a string as the given type; otherwise returns the input wrapped in SimpleError
+pub fn parse_as<T: DeserializeOwned>(input: &str) -> Result<T, SimpleError> {
+    match serde_json::from_str(input) {
+        Ok(res) => Ok(res),
+        Err(_) => Err(SimpleError::new(input)),
+    }
+}
+
+pub fn read_stdin() -> Result<String, SimpleError> {
+    let mut buffer = String::new();
+    try_with!(io::stdin().read_to_string(&mut buffer), "");
+    Ok(buffer)
+}
+
+
+/// Tries to parse STDIN as the given type; otherwise returns the input wrapped in SimpleError
+pub fn parse_stdin_as<T: DeserializeOwned>() -> Result<T, SimpleError> {
+    let mut buffer = String::new();
+    match io::stdin().read_to_string(&mut buffer) {
+        Err(_) => Err(SimpleError::new("")),
+        Ok(_) => parse_as::<T>(&buffer),
+    }
+}
+
+pub fn choose_or<T>(input: Option<T>, default: T) -> T {
+    if let Some(x) = input {
+        x
+    } else {
+        default
+    }
+}
+
+pub fn choose_from<T: Clone>(cli: &Option<T>, input: &Option<T>) -> Option<T> {
```

Comment:
iiuc, this just [Option::or](https://doc.rust-lang.org/std/option/enum.Option.html#method.or) with a clone at the end? So I wonder if you might just use `foo.or(bar).map(|x| x.clone)`, or at least implement this in terms of that? Personall, id be in favor of the first option, since it's nice for readers to be able to think in terms of standard library functions when possible, rather than having to lookup idiosyncratic, shallow wrappers around them.

If the latter, then this function should perhaps be renamed to `option_or_clone` or something, so it's clear how this is related to `Option::or`, with the addition of a clone.

Reply:
very nice! I was simply unaware of that function; replaced my homegrown one with this everywhere...